Threadmonitor

oalfaro2 opened the pull request:
Added new thread monitoring class

Kakon24 committed:
Merge pull request #20 from Kakon24/dev

Version 1.0

Michael-Reefe committed:
Guiding w/o focusing or calibration

Michael-Reefe committed:
Calibration w/o focusing or guiding

Michael-Reefe committed:
Description fix for --nocalibration arg

Michael-Reefe committed:
Removed test driver that should not have been
included in this branch

Michael-Reefe committed:
Guiding centroid algorithm improvements

Michael-Reefe committed:
Dome shutter fix

Kakon24 committed:
Merge pull request #23 from Kakon24/feature/guiding

Feature/guiding

Kakon24 committed:
Merge pull request #24 from Kakon24/fix/dome_shutter

Dome shutter fix

Michael-Reefe committed:
Weather fix

Michael-Reefe committed:
Lowered stop time

Michael-Reefe committed:
Another small fix

Michael-Reefe committed:
Buffer time

Michael-Reefe committed:
More bugfixes for slewing & tracking

Michael-Reefe committed:
Added extra weather check b/w start time check &
initial slew

Kakon24 committed:
Merge pull request #28 from Kakon24/fix/weather

Weather fix

Michael-Reefe committed:
Added/fixed compatibility for different folders
i.e. save to a different folder for each ticket

Michael-Reefe committed:
Fixed

Michael-Reefe committed:
Initial time check before unpark/open shutter

Michael-Reefe committed:
Initial time check before unpark/open shutter

Michael-Reefe committed:
Repeated flat collection fix

Michael-Reefe committed:
Updated with fixes from other branches

Michael-Reefe committed:
Undeleted guider for merging purposes only

Michael-Reefe committed:
Updated guiding files for merging purposes with ops

Michael-Reefe committed:
Adding changes from other fix branches for merging

Michael-Reefe committed:
Minor beginning calibration change

Michael-Reefe committed:
Merge branch 'OPS' into feature/calibration

Michael-Reefe committed:
Fixing accidental removals in merge commimt

Michael-Reefe committed:
Small change to detection of if flats/darks have
already been collected

Michael-Reefe committed:
End time checks during weather shutdown

Michael-Reefe committed:
Fixed last commit

Michael-Reefe committed:
One more small fix

Michael-Reefe committed:
Fixes from folder branches added

Michael-Reefe committed:
Added print statement

Michael-Reefe committed:
Merge branch 'OPS' into fix/folders

Michael-Reefe committed:
A few extra fixes from the folder branches

Michael-Reefe committed:
Simplified weather procedures

Michael-Reefe committed:
Merge branch 'fix/folders' of https://github.com/Kakon24/-OmegaLambda into fix/folders

Michael-Reefe committed:
Adding in fixes from other branches

Michael-Reefe committed:
More nitpicky things

Michael-Reefe committed:
Weather.com fix

Michael-Reefe committed:
Weather.com fix

Michael-Reefe committed:
Fixed changes Shawn suggetsed

Michael-Reefe committed:
Updated w/ fixes from guiding branch

Michael-Reefe committed:
Updated w/ fixes from guiding branch

Michael-Reefe committed:
Shawn's fixes

Michael-Reefe committed:
Shawn's fixes from other branches

Michael-Reefe committed:
Simplified some booleans

Kakon24 committed:
Merge pull request #30 from Kakon24/fix/guiding

Guiding centroid algorithm improvements

Kakon24 committed:
Merge pull request #31 from Kakon24/fix/folders

Fix/folders

Michael-Reefe committed:
Print to logging

Michael-Reefe committed:
An important plus sign

Michael-Reefe committed:
Merge branch 'OPS' into feature/calibration

Kakon24 committed:
Merge pull request #34 from Kakon24/feature/calibration

Feature/calibration

Michael-Reefe committed:
Merge branch 'feature/focusing' into for_merging_into_ops

Michael-Reefe committed:
Fixing mistakes from merge

Michael-Reefe committed:
Fixing another merge mistake

Michael-Reefe committed:
One final fix

Michael-Reefe committed:
removed input prompt

Michael-Reefe committed:
telescope pulse guide fix

Michael-Reefe committed:
small fix

Michael-Reefe committed:
small temperature fix

Commit:
Created new file thread_monitor to run side by side with main code

Michael-Reefe committed:
Guider perpetual waiting fix

oalfaro2 committed:
Delete -omegalambda.iml

oalfaro2 committed:
Delete misc.xml

oalfaro2 committed:
Delete modules.xml

oalfaro2 committed:
Delete vcs.xml

oalfaro2 committed:
Delete profiles_settings.xml

oalfaro2 committed:
Created .gitignore for pycharm on Mac

Commit:
Added import statements to threadmonitor

Commit:
Merge branch 'threadmonitor' of https://github.com/Kakon24/-omegalambda into threadmonitor
Created .gitignore file on github

Commit:
fixing .gitignore

Michael-Reefe committed:
Guiding improvement

Michael-Reefe committed:
small fix to last addition

Michael-Reefe committed:
another fix

Michael-Reefe committed:
another another fix

Commit:
Added restart clauses for threads, needs testing on ops

Michael-Reefe committed:
Accidentally added a bunch of files, made some changes to the
threadmonitor, WIP

oalfaro2 committed:
Delete __main__.cpython-37.pyc

oalfaro2 committed:
Delete __main__.cpython-38.pyc

oalfaro2 committed:
Delete CameraClass.py

oalfaro2 committed:
Delete camera.py

oalfaro2 committed:
Delete commport.py

oalfaro2 committed:
Delete telescope.py

oalfaro2 committed:
Delete omegalambda_log.log

oalfaro2 committed:
Delete omegalambda_log.log.1

oalfaro2 committed:
Delete omegalambda_log.log.2

oalfaro2 committed:
Delete omegalambda_log.log.3

oalfaro2 committed:
Delete Driver Name Fetcher [02-24-20].cpython-37.pyc

oalfaro2 committed:
Delete DriverName.cpython-37.pyc

oalfaro2 committed:
Delete 02-17-20.txt

oalfaro2 committed:
Delete 02-24-20.txt

oalfaro2 committed:
Delete 03-02-20.txt

oalfaro2 committed:
Delete 06-04-20.txt

oalfaro2 committed:
Delete DriverName.py

oalfaro2 committed:
Delete omegalambda_log.log.4

oalfaro2 committed:
Delete omegalambda_log.log.5

oalfaro2 committed:
Delete omegalambda_log.log.6

oalfaro2 committed:
Delete omegalambda_log.log.7

oalfaro2 committed:
Delete omegalambda_log.log.8

oalfaro2 committed:
Delete omegalambda_log.log.9

oalfaro2 committed:
Delete __init__.cpython-37.pyc

oalfaro2 committed:
Delete __init__.cpython-38.pyc

oalfaro2 committed:
Delete logger.cpython-37.pyc

oalfaro2 committed:
Delete logger.cpython-38.pyc

Michael-Reefe committed:
Added restart clause to observation_run using dicts

Michael-Reefe committed:
Merge branch 'threadmonitor' of https://github.com/kakon24/-omegalambda into threadmonitor

Michael-Reefe committed:
Moved restart clauses to observation_run, created dicts for threads

Commit:
Added restart checks throughout observation_run
Added descriptions for new functions
needs testing on OPS

Kakon24 committed:
Merge pull request #40 from Kakon24/for_merging_into_ops

For merging into ops

Kakon24 commented:
Merge conflicts in observation_run.py

Kakon24 commented:
Misclicked close without comment so I reopened PR:

Fix merge conflicts with observation_run.py

Kakon24 commented on .gitignore:
```diff
@@ -0,0 +1,9 @@
+.idea
```
Use this boiler plate .gitignore then add to it if needed.
https://github.com/github/gitignore/blob/master/Python.gitignore

Kakon24 reviewed (changes requested):
Remove all .pyc, .log, .png files.  Observation ticket .jsons should be local only and not added to any remote repository.  Weather statuses as well.  Remove these from the PR and update the .gitignore to add these files so it doesn't happen in the future.

oalfaro2 committed:
Merge branch 'dev' into threadmonitor

oalfaro2 committed:
Update .gitignore

Michael-Reefe committed:
Added more threadchecks in Observation_run
Needs more testing

Michael-Reefe committed:
Merge branch 'threadmonitor' of https://github.com/kakon24/-omegalambda into threadmonitor

Michael-Reefe committed:
Small change to allow observing if focus connection fails

Michael-Reefe committed:
Merge branch 'dev' into focus_connect_fix

Michael-Reefe committed:
weather fixes

Michael-Reefe committed:
Merge branch 'dev' into fix/weather2

Michael-Reefe committed:
Merge branch 'focus_connect_fix' into fix/weather2

Michael-Reefe committed:
Merge branch 'fix/weather2' of https://github.com/Kakon24/-OmegaLambda into fix/weather2

Michael-Reefe committed:
fixed weather file saving issue

Michael-Reefe committed:
fixed small edge case

Michael-Reefe committed:
save directory year fixes

Michael-Reefe committed:
changed year folder from current time to ticket start time

Michael-Reefe commented on main/controller/thread_monitor.py:
```diff
@@ -0,0 +1,52 @@
+        self.crashed = []
+        super(Monitor, self).__init__(name='Monitor')
+
+    def monitor_run(self, threadlist):
```
this method needs to be named just "run" to overwrite the parent class's run method, otherwise the threading won't work properly.

Michael-Reefe replied:
Consequently, the run method also can't take any arguments as far as I'm aware (other than self, of course), so I would make threadlist a self attribute instead that is created via an argument an init

Michael-Reefe commented on main/controller/thread_monitor.py:
```diff
@@ -0,0 +1,52 @@
+        logging.info('Beginning thread monitoring')
+        while self.runthemonitor:
+            for th_name in threadlist:
+                if not threadlist[th_name].is_alive():
```
is threadlist actually a list, or a dictionary or something?  if it's a normal list, this kind of indexing won't work, in this case if I understand what you're doing, I think you'd just want to do th_name.is_alive() since th_name is already an item in threadlist.

Edit:  I saw later that you are actually passing in a dictionary, which is fine, but in that case I think it'd be more clear if you made the for loop as "for th_name in threadlist.keys()" or something, so it's clear that th_name is actually just a string with the name of the thread, and calling threadlist[th_name] is indexing the dictionary.  Also update the docstring to reflect this.

Michael-Reefe commented on main/controller/thread_monitor.py:
```diff
@@ -0,0 +1,52 @@
+                        logging.error('{} thread has raised an exception'.format(threadlist[th_name].name))
+                    self.threadcrash.set()
+
+    def crash_return(self):
```
I don't think you really need a method for this, you could just call Monitor.crashed to get the self.crashed list

Michael-Reefe commented on main/observing/observation_run.py:
```diff
@@ -114,6 +123,13 @@ def everything_ok(self):
 
+        self.monitor.monitor_run(self.th_dict)
+
+        print('---Checking Threads--- LINE 128')
```
remember to use logging statements, never normal prints for this project

Michael-Reefe commented on main/observing/observation_run.py:
```diff
@@ -114,6 +123,13 @@ def everything_ok(self):
         if message:
             logging.error('Hardware connection timeout: {}'.format(message))
 
+        self.monitor.monitor_run(self.th_dict)
```
instead of calling monitor run, you will want to call self.monitor.start(), which will call self.monitor.run() in the background on the proper thread; the way you're doing it now, it will be called on the main thread, so all progress elsewhere will be stopped indefinitely

Michael-Reefe commented on main/observing/observation_run.py:
```diff
@@ -630,3 +673,58 @@ def _shutdown_procedure(self, calibration, cooler=True):
+        '''
+        #print('---In thread Check')
+        threadlist = self.monitor.crash_return()
+        if self.monitor.threadcrash == True:
```
threadcrash is a threading event, not a boolean, so I don't think this will work; threading events are mainly used for wait() statements to synchronize between threads, not send true/false information, but you could instead do something like "if self.monitor.threadcrash.isSet()"

Michael-Reefe commented on main/observing/observation_run.py:
```diff
@@ -630,3 +673,58 @@ def _shutdown_procedure(self, calibration, cooler=True):
+        threadlist = self.monitor.crash_return()
+        if self.monitor.threadcrash == True:
+            #print('Threads have crashed')
+            threadlist = self.monitor.crash_return()
```
why do you grab this twice?

Michael-Reefe commented on main/observing/observation_run.py:
```diff
@@ -630,3 +673,58 @@ def _shutdown_procedure(self, calibration, cooler=True):
+            #print('Threads have crashed')
+            threadlist = self.monitor.crash_return()
+            for thname in threadlist:
+                self.threadrestart(thname)
```
make sure your method calls match the method names?  here you call "threadrestart", but I see it's defined as just "restart"

Michael-Reefe commented on main/observing/observation_run.py:
```diff
@@ -630,3 +673,58 @@ def _shutdown_procedure(self, calibration, cooler=True):
+            Handle of the original thread to restart
+        '''
+
+        for x in self.th_dict:
```
self.th_dict.keys()

Michael-Reefe commented on main/observing/observation_run.py:
```diff
@@ -630,3 +673,58 @@ def _shutdown_procedure(self, calibration, cooler=True):
+        '''
+
+        for x in self.th_dict:
+            print('In thread restart')
```
why don't you just check if thname == each of these strings?  the whole for loop here is pointless

Michael-Reefe replied:
better yet, you could just do self.th_dict[thname] = Camera() if thname == 'camera' else Telescope() if thname == 'telescope' else ....
and then just self.th_dict[thname].start()

Michael-Reefe commented on main/observing/observation_run.py:
```diff
@@ -114,6 +123,13 @@ def everything_ok(self):
+        print('---Checking Threads--- LINE 128')
+        self.threadcheck()
+
+        self.threadcheck()
```
I'm not sure if just scattering these about everywhere in observation_run is the best approach here, though I can't immediately think of a better alternative either since the threads do need to be restarted from observation_run itself.  In any case, I don't think you need 2 in a row here.  Maybe just put one after every weather check or something.

Michael-Reefe commented on main/observing/observation_run.py:
```diff
@@ -449,7 +488,10 @@ def take_images(self, name, num, exp_time, _filter, end_time, path, cycle_filter
-            
+
+            self.crash_check('RoboFocus.exe')
+            self.threadcheck()
```
you definitely don't need more than 1 threadcheck in the same take_images loop

Michael-Reefe commented on observation_tickets/Kornephoros.json:
```diff
@@ -0,0 +1,15 @@
+{"type": "observation_ticket",
```
don't include any of these in commits

Michael-Reefe replied:
make sure to remove the compiled python files and other random stuff like weather data too

Michael-Reefe committed:
webhook test commit

oalfaro2 committed:
Fixed url issue with cloud image retrieval

oalfaro2 committed:
Created autofill for ticket creator
Still WIP, needs refining

oalfaro2 committed:
Slightly adjusted ticket creator layout

oalfaro2 committed:
Delete google.csv

oalfaro2 committed:
Delete info_chart.csv

oalfaro2 committed:
Moved username, passwords, urls to .json file

Michael-Reefe committed:
removing user/password for ttf

Michael-Reefe committed:
removed ss

Michael-Reefe committed:
fix restarting guider after weather shutdown

Commit:
Implemented changes by mreefe, needs testing

Commit:
Deleted a bunch of observation tickets
and weather data

Michael-Reefe commented on main/controller/thread_monitor.py:
```diff
@@ -0,0 +1,40 @@
+import time
```
the .log file and __pycache__ files still should not be included in the commit, make sure you remove those too

Michael-Reefe commented on main/observing/observation_run.py:
```diff
@@ -114,6 +123,8 @@ def everything_ok(self):
         if message:
             logging.error('Hardware connection timeout: {}'.format(message))
 
+        self.monitor.start(self.th_dict)
```
I'm pretty sure start (and therefore run) can't take any arguments (not 100% sure), but you can easily fix this by just passing in the th_dict during initialization.  Add it as a parameter to threadmonitor.\_\_init\_\_ and then when you call self.monitor = Monitor(), change to self.monitor = Monitor(th_dict)

Michael-Reefe replied:
wait, why are you starting the threadmonitor within the everything_ok method?  you should not be doing that, everything_ok is called many times throughout an observation run.  It should be started at the same time all the other threads are started, right under the gui (line 89).  [you can't restart threads more than once - you will want to start it once at the beginning of observing and keep it going for the whole run]

Michael-Reefe commented on main/observing/observation_run.py:
```diff
@@ -125,6 +136,8 @@ def everything_ok(self):
                          "a possible re-open.".format(self.config_dict.min_reopen_time))
             time.sleep(self.config_dict.min_reopen_time * 60)
             if self.conditions.sun:
+                print('---Checking Threads--- LINE 142')
```
remember to change these to logging statements; also don't include the line number, as that will change frequently as you make more edits--it's already incorrect

Michael-Reefe commented on main/observing/observation_run.py:
```diff
@@ -630,3 +667,54 @@ def _shutdown_procedure(self, calibration, cooler=True):
+            Handle of the original thread to restart
+        '''
+
+        for x in self.th_dict.keys():
```
why are you looping here?  this will just restart every thread whenever it's called; you don't make reference to the passed in thname at all...shouldn't the if/else statements be asking if thname == 'camera' etc. instead?

Michael-Reefe commented on main/observing/observation_run.py:
```diff
@@ -630,3 +667,54 @@ def _shutdown_procedure(self, calibration, cooler=True):
+        None
+        '''
+        threadlist = self.monitor.crashed()
+        if self.monitor.threadcrash.isSet():
```
I'm not sure if you really need this event at all?  if there are any threads in the threadlist then you already know they have crashed, in this case you can just do something like:  "if threadlist:" then the for loop.

Michael-Reefe commented on main/controller/thread_monitor.py:
```diff
@@ -0,0 +1,40 @@
+        while self.run_th_monitor:
+            for th_name in threadlist.keys():
+                if not threadlist[th_name].is_alive():
+                    if not threadlist[th_name].name in self.crashed:
```
you can make this an "and" statement with the if above

one more thing:  this loop will go incredibly fast and hog a lot of CPU space while it's running...you'd better add a time.sleep for 5-15 minutes or something so it's not checking 1000000 times every second

Michael-Reefe commented on main/controller/thread_monitor.py:
```diff
@@ -0,0 +1,40 @@
+                    if not threadlist[th_name].name in self.crashed:
+                        self.crashed.append(threadlist[th_name].name)
+                        logging.error('{} thread has raised an exception'.format(threadlist[th_name].name))
+                    self.threadcrash.set()
```
mentioned in a later comment: I'm not sure if you really need this event

Michael-Reefe commented on main/observing/observation_run.py:
```diff
@@ -630,3 +667,54 @@ def _shutdown_procedure(self, calibration, cooler=True):
+            elif x == 'conditions':
+                self.conditions = Conditions()
+                self.conditions.start()
+            elif x == 'guider':
```
also, I'd make sure at the end of this method that you remove the appropriate thread from the crashed threadlist, otherwise once it crashes once it will be continuously restarted

Michael-Reefe commented on main/observing/observation_run.py:
```diff
@@ -61,6 +63,7 @@ def __init__(self, observation_request_list, image_directory, shutdown_toggle, c
         self.conditions = Conditions()
         self.focuser = Focuser()
         self.flatlamp = FlatLamp()
+        self.monitor = Monitor()
```
this is very nitpicky, but for organizational purposes I'd put the monitor initailization after the gui (line 72), maybe even with its own comment header saying "Initializes heartbeat monitor" or something

Michael-Reefe commented on main/observing/observation_run.py:
```diff
@@ -125,6 +136,8 @@ def everything_ok(self):
             time.sleep(self.config_dict.min_reopen_time * 60)
             if self.conditions.sun:
+                print('---Checking Threads--- LINE 142')
+                self.threadcheck()
```
you shouldn't be doing a threadcheck here...if you want to do one at the start of everything_ok, that's fine (~ line 129), but it shouldn't be within this if statement.  If anything, you might want to put an additional one in the while loop at line 147 since we may be waiting there for potentially hours.  same deal for the while loop at line 155, I'd add one there too.

Michael-Reefe commented on main/observing/observation_run.py:
```diff
@@ -149,6 +162,9 @@ def everything_ok(self):
                 check = True
                 self._startup_procedure(cooler=cooler)
+
+                self.threadcheck()
```
you don't want a threadcheck here, see above comment

Michael-Reefe commented on main/observing/observation_run.py:
```diff
@@ -176,6 +192,8 @@ def _startup_procedure(self, cooler=True):
         """
         initial_check = self.everything_ok()
+        print('---Checking Threads--- LINE 198')
+        self.threadcheck()
```
assuming you fix the threadchecks in everything_ok, you don't need an additional one here, as you will have just done one within everything_ok

Michael-Reefe commented on main/observing/observation_run.py:
```diff
@@ -229,12 +247,19 @@ def check_start_time(self, ticket):
             current_epoch_milli = time_utils.datetime_to_epoch_milli_converter(current_time)
             start_time_epoch_milli = time_utils.datetime_to_epoch_milli_converter(ticket.start_time)
-            time.sleep((start_time_epoch_milli - current_epoch_milli) / 1000)
+            #time.sleep((start_time_epoch_milli - current_epoch_milli) / 1000)
```
why did you comment this line out?  this is a very important sleep statement, it definitely should NOT be commented out

also you don't need another threadcheck at line 262

Michael-Reefe commented on main/observing/observation_run.py:
```diff
@@ -229,12 +247,19 @@ def check_start_time(self, ticket):
         if ticket.start_time > current_time:
+
+            print('---Checking Threads--- LINE 256')
+            self.threadcheck()
```
you don't need a threadcheck here...remember these large blocks of code may look long, but most of it will be executed within milliseconds, you don't need to spam threadchecks everywhere

Michael-Reefe commented on main/observing/observation_run.py:
```diff
@@ -257,9 +282,14 @@ def observe(self):
             cooler = True
-
+        
+        self.threadcheck()
```
again this is unnecessary, the _startup_procedure method calls everything_ok, which in turn will call threadcheck

Michael-Reefe replied:
this one is still unnecessary

Michael-Reefe commented on main/observing/observation_run.py:
```diff
@@ -257,9 +282,14 @@ def observe(self):
         initial_shutter = self._startup_procedure(cooler=cooler)
+
+        print('---Threadcheck at 295')
+        self.threadcheck()
```
unnecessary

Michael-Reefe commented on main/observing/observation_run.py:
```diff
@@ -321,6 +351,7 @@ def focus_target(self, ticket):
         None.
 
         """
+        self.threadcheck()
```
unnecessary
it's this method where you might want to do something a bit more involved, because the wait() statement at line 369 will be there for quite a while as the startup focus procedure takes ~10 images...you could add a threading event when the startup_focus_procedure is finished (we might even already have one), say it's called focused_done, and make a while loop:
while not self.focus_procedures.focused_done.isSet():
     time.sleep(60)
     self.threadcheck()

Michael-Reefe commented on main/observing/observation_run.py:
```diff
@@ -353,7 +384,9 @@ def run_ticket(self, ticket):
             The total number of images that are specified on the
             observation ticket.
         """
+        self.threadcheck()
```
unnecessary here

Michael-Reefe commented on main/observing/observation_run.py:
```diff
@@ -418,6 +451,7 @@ def take_images(self, name, num, exp_time, _filter, end_time, path, cycle_filter
         i = 0
         while i < num:
             logging.debug('In take_images loop')
+            self.threadcheck()
```
this is unnecessary because we already call everything_ok once every loop

Michael-Reefe commented on main/observing/observation_run.py:
```diff
@@ -482,6 +518,7 @@ def crash_check(self, program):
             False.
 
         """
+        self.threadcheck()
```
unnecessary
you might as well just remove all the calls to threadcheck() outside of the everything_ok() method and just leave it up to everything_ok to call when it's necessary, as those are already fairly spread out throughout the code.  as long as you fix the threadcheck calls within everything_ok, as previously mentioned, it should be just fine

Michael-Reefe commented on resources/weather_status/weather.txt:
```diff
@@ -0,0 +1 @@
+b'<HTML>\r\n<HEAD>\r\n<TITLE>Current Weather Conditions  at Fairfax, Virginia</TITLE>\r\n</HEAD>\r\n<meta http-equiv="refresh" content="150";>\r\n<BODY bgcolor=PaleGoldenrod>\r\n\r\n<table border="1" bordercolordark="#000000" bordercolorlight="#000000"  cellpadding=2 cellspacing=2>\r\n<tr height="30">\r\n <td align=center bordercolor="#000000" colspan="2"\r\n style="border: thin solid rgb(0,0,0)" ...
```
this should also not be included in the commit

Michael-Reefe committed:
added telescope coordinate check for hour angles > 9h

oalfaro2 committed:
Modified where the RA and Dec come from
Chrome does not like authentification in the URL, switched sites

oalfaro2 committed:
Delete url_config.json

Michael-Reefe committed:
added a few extra waits in case dome pauses while syncing

oalfaro2 committed:
fixed hour angle calculations

Michael-Reefe committed:
fixed ha calculations (again)

Michael-Reefe committed:
added saving focus data to txt

Michael-Reefe committed:
added formatting to focus data filesaving

Michael-Reefe committed:
added shutdown between tickets if there is a long
wait time (due to weather concerns)

Michael-Reefe committed:
shortened threshold for time between tickets
(for safety/paranoia)

Kakon24 committed:
Merge pull request #48 from Kakon24/ticket_creator

Ticket creator

Michael-Reefe committed:
fixed focusing threading event

oalfaro2 committed:
Threadmonitor now working, needs real test
Tracks number of times threads restart, prints
at the end of the night

oalfaro2 committed:
Delete __main__.cpython-38.pyc

oalfaro2 committed:
Delete omegalambda_log.log

oalfaro2 committed:
Delete radar-img0001.png

oalfaro2 committed:
Delete radar-img0002.png

oalfaro2 committed:
Delete cloud-img.gif

oalfaro2 committed:
Delete radar-img0003.png

oalfaro2 committed:
Delete radar-img0004.png

Michael-Reefe commented on main/controller/thread_monitor.py:
```diff
@@ -0,0 +1,47 @@
+        '''
+        logging.info('Beginning thread monitoring')
+        while self.run_th_monitor:
+            for th_name in threadlist.keys():
```
you took out the threadlist argument (good), but now threadlist is undefined.
what I meant for you to do was make threadlist an instance attribute, so you need to rewrite your init method like so:

def \_\_init\_\_(self, th_dict):
    self.threadlist = th_dict
    .... [all the stuff you already have]
    
 then in the run method you can just call self.threadlist
 this also means that when you instantiate the monitor in observation run, you need to do self.monitor = Monitor(th_dict)

Michael-Reefe commented on main/observing/observation_run.py:
```diff
@@ -84,6 +87,13 @@ def __init__(self, observation_request_list, image_directory, shutdown_toggle, c
+        self.th_dict = {'camera': self.camera, 'telescope': self.telescope,
+                        'dome': self.dome, 'focuser': self.focuser, 'flatlamp': self.flatlamp,
+                        'Conditions-Th': self.conditions, 'guider': self.guider,
+                        }
```
you are missing some threads here....there is also focus_procedures and GUI threads
also why do you use the official 'Conditions-Th' name for that one and none of the other ones?  all the threads are named similarly like 'Camera-Th', 'FlatLamp-Th', etc.

Michael-Reefe commented on main/observing/observation_run.py:
```diff
@@ -229,12 +245,18 @@ def check_start_time(self, ticket):
         None.
         """
         current_time = datetime.datetime.now(self.tz)
+        time.sleep(10)
```
why did you add a sleep time here?

Michael-Reefe commented on main/controller/thread_monitor.py:
```diff
@@ -0,0 +1,47 @@
+                    self.threadcrash.set()
+
+
+    def pass_dict(self, th_dict):
```
this method is unneeded if you do what I explain above

Michael-Reefe commented on main/observing/observation_run.py:
```diff
@@ -630,3 +662,62 @@ def _shutdown_procedure(self, calibration, cooler=True):
+            for thname in threadlist:
+                self.restart(thname)
+        else:
+            logging.info('All threads OK')
```
make this a debug statement

Michael-Reefe commented on main/observing/observation_run.py:
```diff
@@ -630,3 +662,62 @@ def _shutdown_procedure(self, calibration, cooler=True):
+        thname : Handle
+            Handle of the original thread to restart
+        '''
+        logging.info('Restarting thread {}'.format(thname))
```
this should probably be a logging.error

Michael-Reefe commented on resources/weather_status/Radar coordinates.txt:
```diff
@@ -0,0 +1,35 @@
+Bottom-Right: xyz=(292,392,10)
```
don't include this in the commit

Michael-Reefe commented on resources/weather_status/radar.txt:
```diff
@@ -0,0 +1,234 @@
+<!doctype html>
```
don't include this in the commit either

Michael-Reefe commented on resources/weather_status/weather.txt:
```diff
@@ -0,0 +1 @@
+b'<HTML>\r\n<HEAD>\r\n<TITLE>Current Weather Conditions  at Fairfax, Virginia</TITLE>\r\n</HEAD>\r\n<meta http-equiv="refresh" content="150";>\r\n<BODY bgcolor=PaleGoldenrod>\r\n\r\n<table border="1" bordercolordark="#000000" bordercolorlight="#000000"  cellpadding=2 cellspacing=2>\r\n<tr height="30">\r\n <td align=center bordercolor="#000000" colspan="2"\r\n style="border: thin solid rgb(0,0,0)" ...
```
don't include in commit

Michael-Reefe reviewed (changes requested):
A lot of my comments from the last review still apply (mostly removing unnecessary self.threadcheck() calls).  Please make sure to look through all of the comments carefully.  [Go to "Files Changed" to see the file structure and exactly where I've commented on each file]

oalfaro2 committed:
Implemented changes suggested by mreefe
Cleaned up number of threadchecks

oalfaro2 committed:
Merge branch 'threadmonitor' of https://github.com/kakon24/-omegalambda into threadmonitor

oalfaro2 committed:
Delete Radar coordinates.txt

oalfaro2 committed:
Delete radar.txt

oalfaro2 committed:
Delete weather.txt

Michael-Reefe commented on main/observing/observation_run.py:
```diff
@@ -61,6 +63,7 @@ def __init__(self, observation_request_list, image_directory, shutdown_toggle, c
         self.conditions = Conditions()
         self.focuser = Focuser()
         self.flatlamp = FlatLamp()
+        self.monitor = Monitor()
```
you correctly added the th_dict as an init parameter for monitor in th_monitor.py, but now you need to actually pass it in when you instantiate it, so you need to define self.th_dict _before_ this and then do self.monitor = Monitor(self.th_dict)

Michael-Reefe replied:
don't forget about this

Michael-Reefe commented on main/controller/thread_monitor.py:
```diff
@@ -0,0 +1,43 @@
+                if not self.threadlist[th_name].is_alive():
+                    if not self.threadlist[th_name].name in self.crashed:
+                        self.crashed.append(self.threadlist[th_name].name)
+                        logging.error('{} thread has raised an exception'.format(self.threadlist[th_name].name))
```
you still need to add a time.sleep here

Michael-Reefe replied:
still need this

Michael-Reefe commented on main/observing/observation_run.py:
```diff
@@ -335,6 +352,9 @@ def focus_target(self, ticket):
             focus_exposure = 30
         self.focus_procedures.onThread(self.focus_procedures.startup_focus_procedure, focus_exposure,
                                        self.filterwheel_dict[focus_filter], self.image_directories[ticket])
+        while not self.focus_procedures.focus_isfinished.isSet():
```
we already have an event for this, self.focus_procedures.focused
just use that one

Michael-Reefe commented on main/controller/focuser_procedures.py:
```diff
@@ -55,6 +55,7 @@ def __init__(self, focus_obj, camera_obj, conditions_obj):
        
         self.focused = threading.Event()
         self.continuous_focusing = threading.Event()
+        self.focus_isfinished = threading.Event()
```
self.focused is already the same thing, just use that one
(plus you never have anywhere where you 'set' this event so it wouldn't work anyways)

Michael-Reefe commented on main/observing/observation_run.py:
```diff
@@ -235,6 +249,7 @@ def check_start_time(self, ticket):
             current_epoch_milli = time_utils.datetime_to_epoch_milli_converter(current_time)
             start_time_epoch_milli = time_utils.datetime_to_epoch_milli_converter(ticket.start_time)
             time.sleep((start_time_epoch_milli - current_epoch_milli) / 1000)
+        time.sleep(5)
```
what is this extra sleep for?

Michael-Reefe commented on main/observing/observation_run.py:
```diff
@@ -301,7 +318,7 @@ def observe(self):
             logging.info("{} out of {} exposures were taken for {}.  Moving on to next target.".format(taken, total,
                                                                                                        ticket.name))
-
+            self.focus_procedures.focus_isfinished.clear()
```
see other comments, you don't need this extra event

Michael-Reefe commented on main/observing/observation_run.py:
```diff
@@ -614,6 +637,7 @@ def _shutdown_procedure(self, calibration, cooler=True):
         """
         logging.info("Shutting down observatory.")
         time.sleep(5)
+        self.monitor.run_th_monitor = False
```
Do NOT add this here.  _shutdown_procedure is called during observations when we temporarily close up due to weather, and you don't want to stop the monitor at those times.  Place this at the end of stop_threads() which is only called at the very very end.

Michael-Reefe commented on main/observing/observation_run.py:
```diff
@@ -630,3 +654,70 @@ def _shutdown_procedure(self, calibration, cooler=True):
+        thname : Handle
+            Handle of the original thread to restart
+        '''
+        logging.error('Restarting thread {}'.format(thname))
```
this method looks good now

Michael-Reefe commented on main/observing/observation_run.py:
```diff
@@ -140,6 +152,7 @@ def everything_ok(self):
 
             else:
```
don't forget to add a threadcheck under the "while current_time < sunset_time" loop (currently line 146)

Michael-Reefe reviewed (changes requested):
Still not quite ready, see comments

Michael-Reefe committed:
guider star selection tweaks

oalfaro2 committed:
Fixed threadmonitor with mchaels input

Michael-Reefe commented on main/observing/observation_run.py:
```diff
@@ -630,3 +652,70 @@ def _shutdown_procedure(self, calibration, cooler=True):
             self.camera.onThread(self.camera.cooler_set, False)
+        logging.info('Stopping thread monitoring')
+        self.monitor.run_th_monitor = False
+        logging.debug('Number of thread restarts: {}'.format(self.monitor.n_restarts))
```
I saw you added this to the stop_threads method but you still need to get rid of this

oalfaro2 committed:
Merge branch 'ticket_creator' into threadmonitor
Added further fixes to thread monitoring

oalfaro2 committed:
Bugfixes from merge

oalfaro2 committed:
More bugfixes

Michael-Reefe committed:
implemented easier arctan2

Michael-Reefe committed:
weather reopen time check

Michael-Reefe committed:
fixed domain for guider angles

Michael-Reefe committed:
made guider angle formula more compact

Michael-Reefe committed:
more cleaning up guider angles

oalfaro2 committed:
Small changes
Removed x thread is alive messages
moved threadmonitor shutdown to beginning of shutdown process

oalfaro2 committed:
Merge branch 'focus_thread_fix' into threadmonitor

oalfaro2 committed:
Bug fixes for thread restarting
changed the way threads are checked against the restart threadnames

oalfaro2 committed:
Minor change from threadlist[name].name to threadlist[name]

oalfaro2 committed:
Tiny change
